lib/src/annotate.rs

```diff
@@ -0,0 +1,394 @@
+        .union(
+            &RevsetExpression::commit(starting_commit_id.clone())
+                .ancestors()
+                .filtered(predicate),
+        )
+        .evaluate_programmatic(repo)
+        .map_err(|e| match e {
+            RevsetEvaluationError::StoreError(backend_error) => backend_error,
+            RevsetEvaluationError::Other(_) => {
+                panic!("Unable to evaluate internal revset")
+            }
+        })?;
+    let mut commit_line_map = get_initial_commit_line_map(starting_commit_id, num_lines);
+    let mut original_line_map = HashMap::new();
+
+    for (cid, edge_list) in revset.iter_graph() {
+        let current_commit_line_map = commit_line_map.remove(&cid);
+        if let Some(current_commit_line_map) = current_commit_line_map {
+            let current_commit = repo.store().get_commit(&cid)?;
+            let (new_original_lines, new_commit_lines) = process_commit(
+                repo,
+                file_name,
+                &current_commit,
+                &mut file_cache,
+                current_commit_line_map,
+                &edge_list,
+            )?;
+            original_line_map.extend(new_original_lines);
+            commit_line_map.extend(new_commit_lines);
+            if original_line_map.len() >= num_lines {
+                break;
+            }
+        }
+    }
+    Ok(original_line_map)
+}
+
+/// For a given commit, for each parent, we compare the version in the parent
+/// tree with the current version, updating the mappings for any lines in
+/// common. If the parent doesn't have the file, we skip it.
+/// After iterating through all the parents, any leftover lines unmapped means
+/// that those lines are original in the current commit. In that case,
+/// original_line_map is updated for the leftover lines.
+/// We return the lines that are the same in the child commit and
+/// any parent. Namely, if line x is found in parent Y, we record the mapping
+/// that parent Y has line x. The line mappings for all parents are returned
+/// along with any lines originated in the current commit
+fn process_commit(
+    repo: &dyn Repo,
+    file_name: &RepoPath,
+    current_commit: &Commit,
+    file_cache: &mut FileCache,
+    mut current_commit_line_map: HashMap<usize, usize>,
+    edges: &Vec<GraphEdge<CommitId>>,
+) -> Result<(OriginalLineMap, CommitLineMap), BackendError> {
+    let mut original_line_map = HashMap::new();
+    let mut commit_line_map: CommitLineMap = HashMap::new();
+    for parent_edge in edges {
+        if parent_edge.edge_type != GraphEdgeType::Missing {
+            let parent_commit = repo.store().get_commit(&parent_edge.target)?;
+            let same_line_map = process_files_in_commits(
+                repo.store(),
+                file_name,
+                file_cache,
+                current_commit,
+                &parent_commit,
+            )?;
+
+            for (current_line_number, parent_line_number) in same_line_map {
```

Comment:
resolve `commit_line_map[parent_commit.id()]` out of the loop. You can use `.entry(parent_id).or_default()`.
https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry

Perhaps, `remap_line()` and `mark_lines_from_original()` can be inlined.

Reply:
I played around with inlining quite a bit and the function just ended up a long spaghetti. In general I prefer smaller functions and breaking it out when things get big. 
I think same thing with entry. I tries to move that inside but the mutable borrows got really messy to the point where I was losing readability (or cloning excessively). 
If you can find a clean way to do what you are suggesting, let me know.

Reply:
Here we can initialize the parent line map first. It's nice that we move hash map lookup out of the loop. (And that's the reason I suggested to `.remove()` the `current_commit_line_map` first.)
```rust
            // Perhaps, the initialization can be combined with load_file_into_cache()??
            // let (parent_contents, parent_commit_line_map) = get_or_load(&mut ..)?;
            let parent_commit_line_map = commit_line_map
                .entry(parent_commit.id().clone())
                .or_default();
            for (current_line_number, parent_line_number) in same_line_map {
                if let Some(original_line_number) =
                    current_commit_line_map.remove(&current_line_number)
                {
                    parent_commit_line_map.insert(parent_line_number, original_line_number);
                }
            }
            // TODO: remove entry if parent_commit_line_map was empty?
            // or return/continue early if same_line_map...peek().is_none()
```

Reply:
thanks, that helped a bit simplify things, uploaded a new version addressing this